src/theory/uf/lambda_lift.cpp

```diff
@@ -0,0 +1,215 @@
+}
+
+TrustNode LambdaLift::ppRewrite(Node node, std::vector<SkolemLemma>& lems)
+{
+  TNode skolem = getSkolemFor(node);
+  if (skolem.isNull())
+  {
+    return TrustNode::null();
+  }
+  d_lambdaMap[skolem] = node;
+  if (!options().uf.ufHoLazyLambdaLift)
+  {
+    TrustNode trn = lift(node);
+    lems.push_back(SkolemLemma(trn, skolem));
+  }
+  // if no proofs, return lemma with no generator
+  if (d_epg == nullptr)
+  {
+    return TrustNode::mkTrustRewrite(node, skolem);
+  }
+  Node eq = node.eqNode(skolem);
+  return d_epg->mkTrustedRewrite(
+      node, skolem, PfRule::MACRO_SR_PRED_INTRO, {eq});
+}
+
+bool LambdaLift::needsLift(TNode skolem) const
+{
+  // must have a lambda, and have not have been lifted already
+  Node l = getLambdaFor(skolem);
+  if (l.isNull())
+  {
+    return false;
+  }
+  return d_lifted.find(skolem) == d_lifted.end();
+}
+
+Node LambdaLift::getLambdaFor(TNode skolem) const
+{
+  NodeNodeMap::const_iterator it = d_lambdaMap.find(skolem);
+  if (it == d_lambdaMap.end())
+  {
+    return Node::null();
+  }
+  return it->second;
+}
+
+Node LambdaLift::getAssertionFor(TNode node)
+{
+  TNode skolem = getSkolemFor(node);
+  if (skolem.isNull())
+  {
+    return Node::null();
+  }
+  Kind k = node.getKind();
+  Node assertion;
+  if (k == LAMBDA)
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    // The new assertion
+    std::vector<Node> children;
+    // bound variable list
+    children.push_back(node[0]);
+    // body
+    std::vector<Node> skolem_app_c;
+    skolem_app_c.push_back(skolem);
+    skolem_app_c.insert(skolem_app_c.end(), node[0].begin(), node[0].end());
+    Node skolem_app = nm->mkNode(APPLY_UF, skolem_app_c);
+    children.push_back(skolem_app.eqNode(node[1]));
+    // axiom defining skolem
+    assertion = nm->mkNode(FORALL, children);
+
+    // Lambda lifting is trivial to justify, hence we don't set a proof
+    // generator here. In particular, replacing the skolem introduced
+    // here with its original lambda ensures the new assertion rewrites
+    // to true.
+    // For example, if (lambda y. t[y]) has skolem k, then this lemma is:
+    //   forall x. k(x)=t[x]
+    // whose witness form rewrites
+    //   forall x. (lambda y. t[y])(x)=t[x] --> forall x. t[x]=t[x] --> true
+  }
+  return assertion;
+}
+
+Node LambdaLift::getSkolemFor(TNode node)
+{
+  Node skolem;
+  Kind k = node.getKind();
+  if (k == LAMBDA)
+  {
+    // if a lambda, do lambda-lifting
+    if (!expr::hasFreeVar(node))
+    {
+      Trace("rtf-proof-debug")
+          << "RemoveTermFormulas::run: make LAMBDA skolem" << std::endl;
+      // Make the skolem to represent the lambda
+      NodeManager* nm = NodeManager::currentNM();
+      SkolemManager* sm = nm->getSkolemManager();
+      skolem = sm->mkPurifySkolem(
+          node,
+          "lambdaF",
+          "a function introduced due to term-level lambda removal");
+    }
+  }
+  else if (k == WITNESS)
```

Comment:
How come `witness` is related to this class `LambdaLift`?

Reply:
Didn't mean to include this, i removed now.